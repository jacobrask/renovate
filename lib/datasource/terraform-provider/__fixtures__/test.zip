version https://git-lfs.github.com/spec/v1
oid sha256:b7871907b3f8222cda1cfb48930d96e89d82adf965da3e934036488be64a475b
size 618
