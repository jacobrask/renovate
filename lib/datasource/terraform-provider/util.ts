export const RepositoryRegex = /^(?:(?<namespace>[^\s]+)\/)?(?<dependency>[^\s]+)$/;
