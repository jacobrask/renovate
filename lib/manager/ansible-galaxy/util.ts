export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /.+\..+/;
