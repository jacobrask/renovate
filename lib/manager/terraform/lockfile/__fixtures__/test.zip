version https://git-lfs.github.com/spec/v1
oid sha256:1721d74eaf36c245148394532dc93854e5568352efad27e57361864e7edfded7
size 473638
