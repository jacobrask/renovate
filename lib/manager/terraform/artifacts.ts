import { logger } from '../../logger';
import { ExecOptions, exec } from '../../util/exec';
import {
  deleteLocalFile,
  getSiblingFileName,
  readLocalFile,
  writeLocalFile,
} from '../../util/fs';
import type {
  UpdateArtifact,
  UpdateArtifactsConfig,
  UpdateArtifactsResult,
} from '../types';

const lockFile = '.terraform.lock.hcl';

async function terraformLockUpdate(
  config: UpdateArtifactsConfig
): Promise<void> {
  // TODO do not hardcode
  const platformArch = ['linux_amd64', 'darwin_amd64', 'windows_amd64'];
  // const tempPathProvider = "/tmp/terraform/providers"
  const tempPathProvider = `${config.cacheDir}/terraform/providers`;

  await deleteLocalFile(lockFile);
  const execOptions: ExecOptions = {
    docker: {
      image: 'renovate/terraform',
    },
  };

  const platformParameters = platformArch
    .map((value) => `-platform=${value}`)
    .join(' ');
  const providerMirrorCMD = `terraform providers mirror ${platformParameters} ${tempPathProvider}`;
  await exec(providerMirrorCMD, execOptions);

  const providerLockCMD = `terraform providers lock -fs-mirror=${tempPathProvider} ${platformParameters}`;
  await exec(providerLockCMD, execOptions);
}

export async function updateArtifacts({
  packageFileName,
  updatedDeps,
  newPackageFileContent,
  config,
}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
  logger.debug(`terraform.updateArtifacts(${packageFileName})`);

  const lockFileName = getSiblingFileName(packageFileName, lockFile);
  const existingLockFileContent = await readLocalFile(lockFileName);
  if (!existingLockFileContent) {
    logger.debug('No .terraform.lock.hcl found');
    return null;
  }
  try {
    await writeLocalFile(packageFileName, newPackageFileContent);
    logger.debug('Updating ' + lockFileName);
    await terraformLockUpdate(config);
    logger.debug(`Returning updated ${lockFile}`);
    const terraformProviderLockContent = await readLocalFile(lockFileName);
    if (existingLockFileContent === terraformProviderLockContent) {
      logger.debug(`${lockFile} is unchanged`);
      return null;
    }
    return [
      {
        file: {
          name: lockFileName,
          contents: terraformProviderLockContent,
        },
      },
    ];
  } catch (err) {
    logger.warn({ err }, 'Failed to update Terraform provider lock file');
    return [
      {
        artifactError: {
          lockFile: lockFileName,
          stderr: err.message,
        },
      },
    ];
  }
}
