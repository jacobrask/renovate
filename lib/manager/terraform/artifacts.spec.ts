import { exec as _exec } from 'child_process';
import _fs from 'fs-extra';
import { join } from 'upath';
import { envMock, mockExecAll } from '../../../test/exec-util';
import { git, mocked } from '../../../test/util';
import { setExecConfig } from '../../util/exec';
import { BinarySource } from '../../util/exec/common';
import * as docker from '../../util/exec/docker';
import * as _env from '../../util/exec/env';
import * as terraform from './artifacts';

jest.mock('fs-extra');
jest.mock('child_process');
jest.mock('../../util/exec/env');
jest.mock('../../util/git');
jest.mock('../../util/http');

const fs: jest.Mocked<typeof _fs> = _fs as any;
const exec: jest.Mock<typeof _exec> = _exec as any;
const env = mocked(_env);

const config = {
  // `join` fixes Windows CI
  localDir: join('/tmp/github/some/repo'),
  dockerUser: 'foobar',
};

describe('.updateArtifacts()', () => {
  beforeEach(async () => {
    jest.resetAllMocks();
    jest.resetModules();

    env.getChildProcessEnv.mockReturnValue(envMock.basic);
    await setExecConfig(config);
    docker.resetPrefetchedImages();
  });
  it('returns null if no .terraform.lock.hcl found', async () => {
    const updatedDeps = ['dep1'];
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'versions.tf',
        updatedDeps,
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
  });
  it('returns null if updatedDeps is empty', async () => {
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'versions.tf',
        updatedDeps: [],
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
  });
  it('returns null if unchanged', async () => {
    fs.readFile.mockResolvedValueOnce('Current .terraform.lock.hcl' as any);
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('Current .terraform.lock.hcl' as any);
    const updatedDeps = ['dep1'];
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'versions.tf',
        updatedDeps,
        newPackageFileContent: '',
        config,
      })
    ).toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });
  it('returns updated .terraform.lock.hcl', async () => {
    git.getFile.mockResolvedValueOnce('Old .terraform.lock.hcl');
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('New .terraform.lock.hcl' as any);
    const updatedDeps = ['dep1'];
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'versions.tf',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated .terraform.lock.hcl for lockfile maintenance', async () => {
    git.getFile.mockResolvedValueOnce('Old .terraform.lock.hcl');
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('New .terraform.lock.hcl' as any);
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'versions.tf',
        updatedDeps: [],
        newPackageFileContent: '{}',
        config: { ...config, updateType: 'lockFileMaintenance' },
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });

  it('returns updated .terraform.lock.hcl with docker', async () => {
    jest.spyOn(docker, 'removeDanglingContainers').mockResolvedValueOnce();
    await setExecConfig({ ...config, binarySource: BinarySource.Docker });
    git.getFile.mockResolvedValueOnce('Old .terraform.lock.hcl');
    const execSnapshots = mockExecAll(exec);
    fs.readFile.mockResolvedValueOnce('New .terraform.lock.hcl' as any);
    const updatedDeps = ['dep1'];
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'provider.tf',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).not.toBeNull();
    expect(execSnapshots).toMatchSnapshot();
  });
  it('catches errors', async () => {
    fs.readFile.mockResolvedValueOnce('Current .terraform.lock.hcl' as any);
    fs.outputFile.mockImplementationOnce(() => {
      throw new Error('not found');
    });
    const updatedDeps = ['dep1'];
    expect(
      await terraform.updateArtifacts({
        packageFileName: 'provider.tf',
        updatedDeps,
        newPackageFileContent: '{}',
        config,
      })
    ).toMatchSnapshot();
  });
});
