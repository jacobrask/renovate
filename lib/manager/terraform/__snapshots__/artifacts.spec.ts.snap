// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.updateArtifacts() catches errors 1`] = `
Array [
  Object {
    "artifactError": Object {
      "lockFile": ".terraform.lock.hcl",
      "stderr": "not found",
    },
  },
]
`;

exports[`.updateArtifacts() returns null if unchanged 1`] = `
Array [
  Object {
    "cmd": "terraform providers mirror -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64 /tmp/terraform/providers",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
  Object {
    "cmd": "terraform providers lock -fs-mirror=/tmp/terraform/providers -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`.updateArtifacts() returns updated .terraform.lock.hcl 1`] = `
Array [
  Object {
    "cmd": "terraform providers mirror -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64 /tmp/terraform/providers",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
  Object {
    "cmd": "terraform providers lock -fs-mirror=/tmp/terraform/providers -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`.updateArtifacts() returns updated .terraform.lock.hcl for lockfile maintenance 1`] = `
Array [
  Object {
    "cmd": "terraform providers mirror -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64 /tmp/terraform/providers",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
  Object {
    "cmd": "terraform providers lock -fs-mirror=/tmp/terraform/providers -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;

exports[`.updateArtifacts() returns updated .terraform.lock.hcl with docker 1`] = `
Array [
  Object {
    "cmd": "docker pull renovate/renovate/terraform",
    "options": Object {
      "encoding": "utf-8",
    },
  },
  Object {
    "cmd": "docker ps --filter name=renovate_renovate_terraform -aq",
    "options": Object {
      "encoding": "utf-8",
    },
  },
  Object {
    "cmd": "docker run --rm --name=renovate_renovate_terraform --label=renovate_child -v \\"/tmp/github/some/repo\\":\\"/tmp/github/some/repo\\" -w \\"/tmp/github/some/repo\\" renovate/renovate/terraform bash -l -c \\"terraform providers mirror -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64 /tmp/terraform/providers\\"",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
  Object {
    "cmd": "docker ps --filter name=renovate_renovate_terraform -aq",
    "options": Object {
      "encoding": "utf-8",
    },
  },
  Object {
    "cmd": "docker run --rm --name=renovate_renovate_terraform --label=renovate_child -v \\"/tmp/github/some/repo\\":\\"/tmp/github/some/repo\\" -w \\"/tmp/github/some/repo\\" renovate/renovate/terraform bash -l -c \\"terraform providers lock -fs-mirror=/tmp/terraform/providers -platform=linux_amd64 -platform=darwin_amd64 -platform=windows_amd64\\"",
    "options": Object {
      "cwd": "/tmp/github/some/repo",
      "encoding": "utf-8",
      "env": Object {
        "HOME": "/home/user",
        "HTTPS_PROXY": "https://example.com",
        "HTTP_PROXY": "http://example.com",
        "LANG": "en_US.UTF-8",
        "LC_ALL": "en_US",
        "NO_PROXY": "localhost",
        "PATH": "/tmp/path",
      },
      "maxBuffer": 10485760,
      "timeout": 900000,
    },
  },
]
`;
